### DISCUSSION TITLE
prometheus scaler: make troubleshooting empty results easier

### DISCUSSION THREAD

**Discussion by or-shachar** [kedacore/keda#2481] [Q&A / Need Help]
Hi,

Today I had a long chat with @JorTurFer about a problem I had with the Prom Scaler.
My query formula was written so that it will always return a positive non-zero value, I tested it on my Prom server, but still, the ExternalMetric returned 0 value, without any error logged in any of Keda components.

Eventually after creating my own testing lab that used parts of the Scaler code - I realized that I was falling into [this if use-case](https://github.com/kedacore/keda/blob/53a81e9f12e9c1833b66e2e0e6fe514db1e9e47e/pkg/scalers/prometheus_scaler.go#L261). Further check showed that my data doesn't get populated immediately and I need to add 10m offest in order to make sure that my query actually catches anything*.

**Two things that surfaced after this confusion:**
1. Does it makes sense to log out some warning of maybe to add an `ExternalMetricValue` a `MetricLabels` that will clarify that the 0 value is not because the value was actually zero but because the length of the data was zero.
2. Does it make sense to add "Data offset" parameter to the scaler so that it will always subtract it from the [`time.Now()`](https://github.com/kedacore/keda/blob/53a81e9f12e9c1833b66e2e0e6fe514db1e9e47e/pkg/scalers/prometheus_scaler.go#L222) it uses?

> ** We use Prometheus Cloudwatch exporter to stream metrics from Cloudwatch to Prometheus (for various reasons, not exclusively for Keda). I know we can use Cloudwatch scaler or SQS scaler but I wanted to save on external API usage. [Apparently, they officially state that data should be queried with 10m offset](https://github.com/prometheus/cloudwatch_exporter#timestamps).

**Comment by JorTurFer** [comment]
So finally you catched the problem! nice to heard it 🔝 🔝 
Related with your proposals, my 50 cents:
1. I'm not an expert with Prometheus, but maybe we could add some debug logs (with debug level) in the scaler for those cases.
2. Related with the offset, I don't have the impact that it could have or if it's possible doing it directly using promql but it's an option. In fact, IDK if it makes sense or nor 🤔
